File: src/ObjectManager/Object.Tes/FilePacks/Records/340-BSGN.Birthsign.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class BSGNRecord : Record
    {
        public override string ToString() => $"BSGN: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField FULL; // Sign name
        public FILEField ICON; // Texture
        public STRVField DESC; // Description
        public List<STRVField> NPCSs = new List<STRVField>(); // TES3: Spell/ability
        public List<FMIDField<Record>> SPLOs = new List<FMIDField<Record>>(); // TES4: (points to a SPEL or LVSP record)

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "FULL":
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "ICON":
                case "TNAM": ICON = new FILEField(r, dataSize); return true;
                case "DESC": DESC = new STRVField(r, dataSize); return true;
                case "SPLO": if (SPLOs == null) SPLOs = new List<FMIDField<Record>>(); SPLOs.Add(new FMIDField<Record>(r, dataSize)); return true;
                case "NPCS": if (NPCSs == null) NPCSs = new List<STRVField>(); NPCSs.Add(new STRVField(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/340-BSGN.Birthsign.swift
//
//  BSGNRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class BSGNRecord: Record {
    public override var description: String { return "BSGN: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var FULL: STRVField! // Sign name
    public var ICON: FILEField! // Texture
    public var DESC: STRVField! // Description
    public var NPCSs: [STRVField]? = nil // TES3: Spell/ability
    public var SPLOs: [FMIDField<Record>]? = nil // TES4: (points to a SPEL or LVSP record)

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "FULL",
             "FNAM": FULL = r.readSTRV(dataSize)
        case "ICON",
             "TNAM": ICON = r.readSTRV(dataSize)
        case "DESC": DESC = r.readSTRV(dataSize)
        case "SPLO": if SPLOs == nil { SPLOs = [FMIDField<Record>]() }; SPLOs?.append(FMIDField<Record>(r, dataSize))
        case "NPCS": if NPCSs == nil { NPCSs = [STRVField]() }; NPCSs?.append(r.readSTRV(dataSize))
        default: return false
        }
        return true
    }
}
